plugins/ida/ida_ko_symbol_mapper.ml

```diff
@@ -0,0 +1,147 @@
+open Core_kernel.Std
+open Bap.Std
+include Self ()
+open Graphlib.Std
+
+let map_pc_to_dest w relocs =
+  let open Option in
+  List.find_map relocs ~f:(fun (pc,dest) -> some_if (w = pc) (dest))
+  >>= fun dest -> return (Bil.int dest)
+
+class bil_ko_symbol_relocator relocs = object(self)
+  inherit Stmt.mapper as super
+
+  (** If there's a BIL jmp with an address to one of our assembly addresses, in
+      relocs, it currently jumps to itself. We need to update it to jump to the
+      extern functions in the lookup table provided by IDA brancher. We
+      substitute it with [w], our address in the table *)
+  method map_jmp e =
+    let open Option in
+    match e with
+    | Bil.Int w ->
+      map_pc_to_dest w relocs
+      |> Option.value ~default:(Bil.int w)
+      |> super#map_jmp
+    | _ -> super#map_jmp e
+end
+
+module Cfg = struct
+  include Graphs.Cfg
+
+  (** No edge update method? Rolling my own... *)
+  let update_edge cfg e e' = cfg |> Edge.remove e |> Edge.insert e'
+
+  (** use the same label (edge which is `Jump, `Cond, etc) *)
+  let update_dst b cfg e =
+    let src = Edge.src e in
+    let lbl = Edge.label e in
+    let e' = Edge.create src b lbl in
+    update_edge cfg e e'
+
+  let update_src b cfg e =
+    let dst = Edge.dst e in
+    let lbl = Edge.label e in
+    let e' = Edge.create b dst lbl in
+    update_edge cfg e e'
+end
+
+let full_insn_of_mem mem arch =
+  Disasm_expert.Basic.with_disasm ~backend:"llvm" arch
+    ~f:(fun dis ->
+        let open Disasm_expert.Basic in
+        dis |> store_kinds |> store_asm |> fun dis' ->
+        insn_of_mem dis' mem)
+
+let remap_block b memory arch relocs =
+  let bil_remapper = new bil_ko_symbol_relocator relocs in
+  let arch = Arch.to_string arch in
+  (* We need a full_insn type to reconstruct an Insn with an updated
+      bil. So re-disassemble from mem to get the full_insn type *)
+  Block.insns b |> List.map ~f:(fun (mem,insn) ->
+      let bil = Insn.bil insn in
+      let bil' = bil_remapper#run bil in
+      match full_insn_of_mem mem arch with
+      | Ok (_,Some x,_) ->
+        (mem,Insn.of_basic ~bil:bil' x)
```

Comment:
Note that this line changes the semantics of the bil attached to the jmp instruction associated with a call site of a relocation. I.e., before it was a relocatable PC-relative jump to itself, and now it jumps to a kernel symbol at an address in a synthetic segment in memory. There seems to be no better way to go about it (we can't insert BIL statements without an associated instruction, as far as I'm aware, and that's probably a good thing).